bank/api/api.go

```diff
@@ -0,0 +1,246 @@
+			return
+		}
+
+		var request domain.UserLoginRequest
+		json.NewDecoder(r.Body).Decode(&request)
+
+		user, err := userStore.Authenticate(request)
+		if err != nil {
+			switch err {
+			case domain.ErrUsernameDoesntExist:
+				errorResponse(w, http.StatusNotFound, "Login Failed", "user", "username", "username_does_not_exist")
+				return
+			case domain.ErrAlreadyLoggedIn:
+				errorResponse(w, http.StatusBadRequest, "Login Failed", "user", "user_status", "already_logged_in")
+				return
+			case domain.ErrWrongPassword:
+				errorResponse(w, http.StatusUnauthorized, "Login Failed", "user", "password", "wrong_password")
+				return
+			default:
+				errorResponse(w, http.StatusInternalServerError, "Login Failed", "error", "unexpected_error", err.Error())
+				return
+			}
+		}
+
+		session, err := sessionStore.StartSession(*user, time.Now().Local().Add(userSessionDuration))
+		if err != nil {
+			errorResponse(w, http.StatusInternalServerError, "Session Initializing Failed", "error", "unexpected_error", err.Error())
+			return
+		}
+
+		*cookie = http.Cookie{Name: "SID", Path: "/", Value: session.SessionID, MaxAge: 300}
+		http.SetCookie(w, cookie)
+	})
+}
+
+func LogoutHandler(sessionStore domain.SessionStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		cookie, err := r.Cookie("SID")
+		if err != nil {
+			cookie = &http.Cookie{Value: "rfBd56ti2SMtY"}
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+
+		session, ok := sessionStore.FindSessionAvailableAt(cookie.Value, time.Now().Local())
+		if !ok {
+			errorResponse(w, http.StatusBadRequest, "Logout Failed", "user", "user_status", "already_logged_out")
+			return
+		}
+
+		err = sessionStore.DeleteSession(session.SessionID)
+		if err != nil {
+			errorResponse(w, http.StatusInternalServerError, "Logout Failed", "error", "unexpected_error", err.Error())
+			return
+		}
+
+		*cookie = http.Cookie{Name: "SID", Path: "/", Value: "deleted", MaxAge: -1}
+		http.SetCookie(w, cookie)
+
+		w.WriteHeader(http.StatusOK)
+	})
+}
+
+func GetUserAccounts(accountStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		session := context.Get(r, "session").(*domain.Session)
+
+		var account domain.Account
+		account.UserID = session.UserID
+
+		result, err := accountStore.GetAccounts(account.UserID)
+		if err != nil {
+			errorResponse(w, http.StatusInternalServerError, "Fetch User Accounts Failed", "error", "unexpected_error", err.Error())
+			return
+		}
+
+		json.NewEncoder(w).Encode(result)
+	})
+}
+
+func NewUserAccount(accountStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		session := context.Get(r, "session").(*domain.Session)
+
+		var account domain.Account
+		json.NewDecoder(r.Body).Decode(&account)
+		account.UserID = session.UserID
+
+		_, err := accountStore.InsertAccount(account)
```

Comment:
InsertAccount(userID, account) will make it more explicit. This will remove and setting of the UserId in the account.